### PR TITLE
Add new metric `d_lambda`

### LINKED ISSUE #799: More image metrics
## 🚀 Feature

just found this package and seems we have not yet all Image metrics :] 
some inspiration came from https://github.com/andrewekhalel/sewar but we aim on own implementation with `torch`

### Alternatives

- ~Mean Squared Error (MSE)~
- ~Root Mean Sqaured Error (RMSE)~
- ~Peak Signal-to-Noise Ratio (PSNR)~
- ~Structural Similarity Index (SSIM)~
- ~Universal Quality Image Index (UQI)~
- ~Multi-scale Structural Similarity Index (MS-SSIM)~
- ~Erreur Relative Globale Adimensionnelle de Synthèse (ERGAS)~
- ~Spatial Correlation Coefficient (SCC)~
- ~Relative Average Spectral Error (RASE)~
- ~Spectral Angle Mapper (SAM)~
- ~Spectral Distortion Index (D_lambda)~
- ~Spatial Distortion Index (D_S)~
- ~Quality with No Reference (QNR)~
- ~Visual Information Fidelity (VIF)~
- ~Block Sensitive - Peak Signal-to-Noise Ratio (PSNR-B)~

### Additional context

any of these metrics would be a separate PR/addition :]

### PR DESCRIPTION
**Pull request by ankitaS11**
## What does this PR do?

Adds New Image Metric - D_Lambda (Spectral Distortion Index)

Fixes #799

## Before submitting

- [ ] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by ankitaS11** [comment]
Note to the reviewers: Working on adding tests and updating doc examples. I will make this PR ready for review very soon.

**Comment by ankitaS11** [comment]
Hi @stancld @SkafteNicki, I have finished implementing `d_lambda`, but I need help with testing this because - 

1. `UQI` is used in `d_lamba` and `sewar` uses uniform filter in `uqi` while we use gaussian filter (the paper mentions gaussian distribution).
2.  Couldn't find any reference implementation for `d_lambda` which matches the formula in paper.

If you can please review the code once and suggest testing strategy, that would be a great help.

File: torchmetrics/image/d_lambda.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        reduction: Literal["elementwise_mean", "sum", "none"] = "elementwise_mean",
+        compute_on_step: bool = True,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+    ) -> None:
+        super().__init__(
+            compute_on_step=compute_on_step,
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+        )
```
Comment (SkafteNicki) [inline torchmetrics/image/d_lambda.py:69]: this has recently been refactored, so all these arguments are now just part of a `**kwargs` arg
```suggestion
        **kwargs
    ) -> None:
        super().__init__(**kwargs)
```

File: torchmetrics/image/d_lambda.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        rank_zero_warn(
+            "Metric `SpectralDistortionIndex` will save all targets and"
+            " predictions in buffer. For large datasets this may lead"
+            " to large memory footprint."
+        )
+
+        self.add_state("ms", default=[], dist_reduce_fx="cat")
+        self.add_state("fused", default=[], dist_reduce_fx="cat")
+        self.reduction = reduction
```
Comment (SkafteNicki) [inline torchmetrics/image/d_lambda.py:78]:
```suggestion
        allowed_reduction = ["elementwise_mean", "sum", "none"]
        if reduction not in allowed_reduction:
            raise ValueError(f"Expected argument `reduction` be one of {allowed_reduction} but got {reduction}")
        self.reduction = reduction
```

File: torchmetrics/functional/image/d_lambda.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+    Args:
+        ms: Low resolution multispectral image
+        fused: High resolution fused image
+        p: Parameter to emphasize large spectral differences (default: 1)
+    """
+
+    if ms.dtype != fused.dtype:
+        raise TypeError(
+            "Expected `ms` and `fused` to have the same data type." f" Got ms: {ms.dtype} and fused: {fused.dtype}."
```
Comment (SkafteNicki) [inline torchmetrics/functional/image/d_lambda.py:38]:
```suggestion
            f"Expected `ms` and `fused` to have the same data type. Got ms: {ms.dtype} and fused: {fused.dtype}."
```

File: torchmetrics/functional/image/d_lambda.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        )
+    _check_same_shape(ms, fused)
+    if len(ms.shape) != 4:
+        raise ValueError(
+            "Expected `ms` and `fused` to have BxCxHxW shape." f" Got ms: {ms.shape} and fused: {fused.shape}."
```
Comment (SkafteNicki) [inline torchmetrics/functional/image/d_lambda.py:43]:
```suggestion
            f"Expected `ms` and `fused` to have BxCxHxW shape. Got ms: {ms.shape} and fused: {fused.shape}."
```

File: torchmetrics/functional/image/d_lambda.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        )
+    if p <= 0:
+        raise ValueError("Expected `p` to be a positive integer." f" Got p: {p}.")
```
Comment (SkafteNicki) [inline torchmetrics/functional/image/d_lambda.py:46]:
```suggestion
        raise ValueError(f"Expected `p` to be a positive integer. Got p: {p}.")
```

**Review by SkafteNicki — COMMENTED** [review]
For testing, what about comparing to the implementation in https://github.com/andrewekhalel/sewar?

Also remember:
* Remember to add entry to `Changelog.md`.
* Remember to add reference in the docs

**Comment by ankitaS11** [comment]
> For testing, what about comparing to the implementation in https://github.com/andrewekhalel/sewar?
> 
> Also remember:
> 
> * Remember to add entry to `Changelog.md`.
> * Remember to add reference in the docs

Thank you for the review @SkafteNicki  :)
`sewar` uses uniform filter in uqi and we have used gaussian filter (as the reference research paper mentions gaussian distribution). So I think we cannot compare our implementation with `sewar`.

**Comment by SkafteNicki** [comment]
@ankitaS11 in that case, we only have the option to write have a similar implementation here:
https://github.com/PyTorchLightning/metrics/blob/master/tests/helpers/non_sklearn_metrics.py
written in `numpy` which we can compare against. It is not optimal, but that is just how it is with some of the newer metrics proposed in research papers.

**Comment by stancld** [comment]
> @ankitaS11 in that case, we only have the option to write have a similar implementation here: https://github.com/PyTorchLightning/metrics/blob/master/tests/helpers/non_sklearn_metrics.py written in `numpy` which we can compare against. It is not optimal, but that is just how it is with some of the newer metrics proposed in research papers.

Or maybe we can adjust the implementation of UQI so that both Gaussian and uniform kernel can be used?

**Comment by ankitaS11** [comment]
> @ankitaS11 in that case, we only have the option to write have a similar implementation here: https://github.com/PyTorchLightning/metrics/blob/master/tests/helpers/non_sklearn_metrics.py written in `numpy` which we can compare against. It is not optimal, but that is just how it is with some of the newer metrics proposed in research papers.

Hi @SkafteNicki, thank you for the suggestion. I have added the tests and also a reference implementation in the file you suggested but it uses UQI from our module so that we have uniformity between the filters we use. All the tests are passing on my system but I am aware that the tests for examples in the docs will fail because I haven't updated the value yet. I will update them in the next commit along with `changelog.md` but it would be great if you and @stancld could review it once.

**Review by SkafteNicki — APPROVED** [review]

**Comment by SkafteNicki** [comment]
Hi @ankitaS11,
Core part of the code looks good to me. I added some remaining things, trying to get it ready for merging.
Thanks :]

**Review by justusschock — APPROVED** [review]

**Review by stancld — APPROVED** [review]
LGTM! :]

cc: @SkafteNicki @justusschock Just one thing - don't we wanna use `preds, target` naming instead of `ms, fused` and just reference to these naming in the docstrings?

**Comment by justusschock** [comment]
@stancld , right. Totally missed that. I think we should do that for consistency :)

File: tests/helpers/non_sklearn_metrics.py [hunk 6]
```diff
@@ -185,3 +188,35 @@ def calibration_error(
...
+
+def d_lambda(ms: np.ndarray, fused: np.ndarray, p: int = 1) -> float:
+    """A NumPy based implementation of Spectral Distortion Index, which uses UQI of TorchMetrics."""
+    fused, ms = torch.from_numpy(fused), torch.from_numpy(ms)
+    # To convert to BxHxWxC
+    # fused, ms = fused[None, ...], ms[None, ...]
+    # breakpoint()
+    # Permute to ensure B x C x H x W (Pillow/NumPy stores in B x H x W x C)
+    fused = torch.permute(fused, (0, 3, 1, 2))
+    ms = torch.permute(ms, (0, 3, 1, 2))
```
Comment (stancld) [inline tests/helpers/non_sklearn_metrics.py:201]:
```suggestion
    fused = fused.permute(0, 3, 1, 2)
    ms = ms.permute(0, 3, 1, 2)
```

File: torchmetrics/functional/image/d_lambda.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    if L == 1:
+        output = torch.pow(diff, (1.0 / p))
+    else:
+        output = torch.pow(1.0 / (L * (L - 1)) * torch.sum(diff), (1.0 / p))
+    return reduce(output, reduction)
+
+
+def spectral_distortion_index(
+    ms: Tensor,
+    fused: Tensor,
```
Comment (stancld) [inline torchmetrics/functional/image/d_lambda.py:98]:
```suggestion
    preds: Tensor,
    target: Tensor,
```
@ankitaS11 Please use `preds, target` naming for variables to stay consistent with the rest of `torchmetrics` package :]

Comment (ankitaS11) [reply torchmetrics/functional/image/d_lambda.py:98]: I've a question regarding this, @stancld - since this metric (`D_LAMBDA`) is a no-reference image metric, so we don't have a `target` per say... Do you still think that renaming it to `preds`/`target` is a good idea?

I believe it's a little confusing, but I also agree that we should stay consistent. So I don't have a strong opinion here, and if you say, I'm happy to rename it for now. Maybe, in the future, we can think of more generic names for all the metrics in the `torchmetrics` API. Just sharing my thoughts :)

cc: @SkafteNicki @justusschock

File: tests/helpers/non_sklearn_metrics.py [hunk 8]
```diff
@@ -185,3 +188,35 @@ def calibration_error(
...
         loss = np.sqrt(max(loss, 0.0))
     return loss
+
+
+def d_lambda(ms: np.ndarray, fused: np.ndarray, p: int = 1) -> float:
+    """A NumPy based implementation of Spectral Distortion Index, which uses UQI of TorchMetrics."""
+    fused, ms = torch.from_numpy(fused), torch.from_numpy(ms)
+    # To convert to BxHxWxC
+    # fused, ms = fused[None, ...], ms[None, ...]
+    # breakpoint()
```
Comment (ankitaS11) [inline tests/helpers/non_sklearn_metrics.py:198]: Removing comments
```suggestion

```

File: torchmetrics/functional/image/d_lambda.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+            - ``'none'``: no reduction will be applied
+
+    Example:
+        >>> import torch
+        >>> _ = torch.manual_seed(42)
+        >>> ms = torch.rand([16, 3, 16, 16])
+        >>> fused = ms * 0.75
+        >>> ms, target = _d_lambda_update(ms, fused)
+        >>> _d_lambda_compute(ms, fused)
+        tensor(3.4769e-08)
```
Comment (ankitaS11) [inline torchmetrics/functional/image/d_lambda.py:72]: Is there any way to skip the doc tests? In `uqi` similar test was passing as we had a constant factor of 0.75 for distortion. (and its a full reference metric so it will give same result without `torch.manual_seed`). Probably `torch.manual_seed` is not solving the problem here as the test is failing everytime with a different expected value.

**Review by ankitaS11 — COMMENTED** [review]
Doc tests are failing, so can someone help with my comment?

File: torchmetrics/functional/image/d_lambda.py (hunk 9 above)
Comment (ankitaS11) [reply torchmetrics/functional/image/d_lambda.py:72]: @Borda, I just checked the CI, it seems to be failing for multiple environments with different values (`Got: some random value`) even after using `manual_seed`.

Eg: https://github.com/PyTorchLightning/metrics/runs/5382467199?check_suite_focus=true,  https://github.com/PyTorchLightning/metrics/runs/5382472779?check_suite_focus=true

Comment (stancld) [reply torchmetrics/functional/image/d_lambda.py]: @ankitaS11 In the previous scenario, it seems like as `ms` and `fused` were too similar, the value of this metric was really close to 0 and there was a discrepancy on the 8th decimal place across some OS/torch versions. I change `fused = ms * 0.75` to `fused = torch.rand([16, 3, 16, 16])` so that the tensors won't be so similar. Hopefully, it looks like everything works fine now :]

**Comment by ankitaS11** [comment]
Hi, I accidentally messed up a few things in my fork, and deleted it :( Sorry, I will create a new one immediately, I hope that's okay.
